### PR TITLE
dvbcsa: add BISS-2 decryption

### PR DESCRIPTION
**Pull request by funman**
The first commits merge the bs and non-bs modes of libdvbcsa into one single pipe.

The last commit is not complete but shows the goal of this PR.

### CODE REVIEW

File: lib/upipe-dvbcsa/upipe_dvbcsa_decrypt.c [hunk 1]
```diff
@@ -101,54 +201,172 @@ static void upipe_dvbcsa_dec_free(struct upipe *upipe)
...
  *
  * @param upipe description structure of the pipe
  * @param uref input buffer to handle
  * @param upump_p reference to the pump that generated the buffer
  */
 static void upipe_dvbcsa_dec_input(struct upipe *upipe,
-                                   struct uref *uref,
-                                   struct upump **upump_p)
+                                      struct uref *uref,
+                                      struct upump **upump_p)
```
Comment (quarium) [inline lib/upipe-dvbcsa/upipe_dvbcsa_decrypt.c]: Indentation

File: lib/upipe-dvbcsa/upipe_dvbcsa_decrypt.c [hunk 2]
```diff
@@ -58,19 +76,92 @@ struct upipe_dvbcsa_dec {
...
+
 /** @hidden */
 UBASE_FROM_TO(upipe_dvbcsa_dec, upipe_dvbcsa_common, common, common);
 
 UPIPE_HELPER_UPIPE(upipe_dvbcsa_dec, upipe, UPIPE_DVBCSA_DEC_SIGNATURE);
-UPIPE_HELPER_UREFCOUNT(upipe_dvbcsa_dec, urefcount, upipe_dvbcsa_dec_free);
-UPIPE_HELPER_VOID(upipe_dvbcsa_dec);
-UPIPE_HELPER_OUTPUT(upipe_dvbcsa_dec, output, flow_def, output_state, requests);
+UPIPE_HELPER_UREFCOUNT(upipe_dvbcsa_dec, urefcount,
+                       upipe_dvbcsa_dec_free);
```
[the remark this answers is not in the document]
Comment (funman) [reply lib/upipe-dvbcsa/upipe_dvbcsa_decrypt.c]: I am not sure what indentation you want to see.
This was literally copy pasting from upipe_dvbcsa_bs_decrypt.c

File: lib/upipe-dvbcsa/upipe_dvbcsa_decrypt.c [hunk 3]
```diff
@@ -253,18 +557,21 @@ static int upipe_dvbcsa_dec_set_key(struct upipe *upipe, const char *key)
...
+                                            va_list args)
 {
     struct upipe_dvbcsa_dec *upipe_dvbcsa_dec =
         upipe_dvbcsa_dec_from_upipe(upipe);
     struct upipe_dvbcsa_common *common =
         upipe_dvbcsa_dec_to_common(upipe_dvbcsa_dec);
-
-    UBASE_HANDLED_RETURN(upipe_dvbcsa_dec_control_output(upipe, command, args));
+    UBASE_HANDLED_RETURN(
+        upipe_dvbcsa_dec_control_output(upipe, command, args));
```
Comment (quarium) [inline lib/upipe-dvbcsa/upipe_dvbcsa_decrypt.c]: Indentation

File: lib/upipe-dvbcsa/upipe_dvbcsa_encrypt.c [hunk 4]
```diff
@@ -162,111 +352,187 @@ static void upipe_dvbcsa_enc_input(struct upipe *upipe,
...
+    upipe_dvbcsa_enc->mapped[current] = uref;
+    upipe_dvbcsa_enc->current++;
+
+    /* hold uref */
+    upipe_dvbcsa_enc_hold_input(upipe, uref);
+    if (unlikely(first)) {
+        /* make sure to send all buffered urefs */
+        upipe_use(upipe);
+        upipe_dvbcsa_enc_wait_upump(upipe, common->latency,
+                                       upipe_dvbcsa_enc_worker);
```
Comment (quarium) [inline lib/upipe-dvbcsa/upipe_dvbcsa_encrypt.c]: Indentation

File: lib/upipe-dvbcsa/upipe_dvbcsa_decrypt.c [hunk 5]
```diff
@@ -180,26 +402,87 @@ static void upipe_dvbcsa_dec_input(struct upipe *upipe,
...
-                   size - ts_header_size);
-    return upipe_dvbcsa_dec_output(upipe, uref, upump_p);
+
+    /* hold uref */
+    upipe_dvbcsa_dec_hold_input(upipe, uref);
+    if (unlikely(first)) {
+        /* make sure to send all buffered urefs */
+        upipe_use(upipe);
+        upipe_dvbcsa_dec_wait_upump(upipe, common->latency,
+                                       upipe_dvbcsa_dec_worker);
```
Comment (quarium) [inline lib/upipe-dvbcsa/upipe_dvbcsa_decrypt.c]: Indentation

File: lib/upipe-dvbcsa/upipe_dvbcsa_decrypt.c [hunk 6]
```diff
@@ -275,14 +582,28 @@ static int upipe_dvbcsa_dec_control(struct upipe *upipe,
...
+/** @internal @This handles a pipe control command and checks the upump
+ * manager.
+ *
+ * @param upipe description structure of the pipe
+ * @param command control command to handle
+ * @param args optional arguments
+ * @return an error code
+ */
+static int upipe_dvbcsa_dec_control(struct upipe *upipe,
+                                       int command, va_list args)
```
Comment (quarium) [inline lib/upipe-dvbcsa/upipe_dvbcsa_decrypt.c]: Indentation

File: lib/upipe-dvbcsa/upipe_dvbcsa_decrypt.c [hunk 7]
```diff
@@ -253,18 +557,21 @@ static int upipe_dvbcsa_dec_set_key(struct upipe *upipe, const char *key)
  * @param args optional arguments
  * @return an error code
  */
-static int upipe_dvbcsa_dec_control(struct upipe *upipe,
-                                    int command,
-                                    va_list args)
+static int upipe_dvbcsa_dec_control_real(struct upipe *upipe,
+                                            int command,
+                                            va_list args)
```
Comment (quarium) [inline lib/upipe-dvbcsa/upipe_dvbcsa_decrypt.c]: Indentation

File: lib/upipe-dvbcsa/upipe_dvbcsa_encrypt.c [hunk 8]
```diff
@@ -23,59 +23,129 @@
...
+/** @hidden */
+static int upipe_dvbcsa_enc_check(struct upipe *upipe,
+                                     struct uref *flow_def);
+
 UPIPE_HELPER_UPIPE(upipe_dvbcsa_enc, upipe, UPIPE_DVBCSA_ENC_SIGNATURE);
-UPIPE_HELPER_UREFCOUNT(upipe_dvbcsa_enc, urefcount, upipe_dvbcsa_enc_free);
-UPIPE_HELPER_VOID(upipe_dvbcsa_enc);
-UPIPE_HELPER_OUTPUT(upipe_dvbcsa_enc, output, flow_def, output_state, requests);
+UPIPE_HELPER_UREFCOUNT(upipe_dvbcsa_enc, urefcount,
+                       upipe_dvbcsa_enc_free);
```
Comment (quarium) [inline lib/upipe-dvbcsa/upipe_dvbcsa_encrypt.c]: Indentation

File: lib/upipe-dvbcsa/upipe_dvbcsa_encrypt.c [hunk 9]
```diff
@@ -23,59 +23,129 @@
...
+UPIPE_HELPER_FLOW(upipe_dvbcsa_enc, NULL);
+UPIPE_HELPER_INPUT(upipe_dvbcsa_enc, urefs, nb_urefs, max_urefs, blockers,
+                   NULL);
```
Comment (quarium) [inline lib/upipe-dvbcsa/upipe_dvbcsa_encrypt.c]: Indentation

File: lib/upipe-dvbcsa/upipe_dvbcsa_encrypt.c [hunk 10]
```diff
@@ -103,15 +181,16 @@ static void upipe_dvbcsa_enc_free(struct upipe *upipe)
...
- * @return an allocated and initialized pipe
+ * @return an allocated and initialized pipe or NULL
  */
 static struct upipe *upipe_dvbcsa_enc_alloc(struct upipe_mgr *mgr,
-                                            struct uprobe *uprobe,
-                                            uint32_t signature,
-                                            va_list args)
+                                               struct uprobe *uprobe,
+                                               uint32_t signature,
+                                               va_list args)
```
Comment (quarium) [inline lib/upipe-dvbcsa/upipe_dvbcsa_encrypt.c]: Indentation

File: lib/upipe-dvbcsa/upipe_dvbcsa_decrypt.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (funman) [reply lib/upipe-dvbcsa/upipe_dvbcsa_decrypt.c]: I guess it's the function declaration above (missing from the chunk on github UI)

[the remark this answers is not in the document]
Comment (quarium) [reply lib/upipe-dvbcsa/upipe_dvbcsa_decrypt.c]: No need to use two lines

File: lib/upipe-dvbcsa/upipe_dvbcsa_decrypt.c [hunk 11]
```diff
@@ -105,25 +205,129 @@ static struct upipe *upipe_dvbcsa_dec_alloc(struct upipe_mgr *mgr,
                                             uint32_t signature,
                                             va_list args)
 {
+    struct uref *flow_def;
     struct upipe *upipe =
-        upipe_dvbcsa_dec_alloc_void(mgr, uprobe, signature, args);
-    if (unlikely(!upipe))
+        upipe_dvbcsa_dec_alloc_flow(mgr, uprobe, signature, args, &flow_def);
```
Comment (quarium) [inline lib/upipe-dvbcsa/upipe_dvbcsa_decrypt.c]: flow_def must be freed

File: lib/upipe-dvbcsa/upipe_dvbcsa_encrypt.c [hunk 12]
```diff
@@ -103,15 +181,16 @@ static void upipe_dvbcsa_enc_free(struct upipe *upipe)
...
  */
 static struct upipe *upipe_dvbcsa_enc_alloc(struct upipe_mgr *mgr,
                                             struct uprobe *uprobe,
                                             uint32_t signature,
                                             va_list args)
 {
+    struct uref *flow_def;
     struct upipe *upipe =
-        upipe_dvbcsa_enc_alloc_void(mgr, uprobe, signature, args);
+        upipe_dvbcsa_enc_alloc_flow(mgr, uprobe, signature, args, &flow_def);
```
Comment (quarium) [inline lib/upipe-dvbcsa/upipe_dvbcsa_encrypt.c]: flow_def must be freed

File: lib/upipe-dvbcsa/upipe_dvbcsa_encrypt.c [hunk 13]
```diff
@@ -292,10 +571,6 @@ static struct upipe_mgr upipe_dvbcsa_enc_mgr = {
     .upipe_control = upipe_dvbcsa_enc_control,
 };
 
-/** @This returns the dvbcsa encryption pipe management structure.
- *
- * @return a pointer to the management structure
- */
 struct upipe_mgr *upipe_dvbcsa_enc_mgr_alloc(void)
```
Comment (quarium) [inline lib/upipe-dvbcsa/upipe_dvbcsa_encrypt.c]: Why did you remove the doc ?

Comment (funman) [reply lib/upipe-dvbcsa/upipe_dvbcsa_encrypt.c]: https://github.com/Upipe/upipe/blob/master/lib/upipe-dvbcsa/upipe_dvbcsa_bs_encrypt.c#L514 I did not

File: lib/upipe-dvbcsa/upipe_dvbcsa_decrypt.c (hunk 11 above)
Comment (funman) [reply lib/upipe-dvbcsa/upipe_dvbcsa_decrypt.c]: Done

File: lib/upipe-dvbcsa/upipe_dvbcsa_encrypt.c (hunk 12 above)
Comment (funman) [reply lib/upipe-dvbcsa/upipe_dvbcsa_encrypt.c]: Done

File: lib/upipe-dvbcsa/upipe_dvbcsa_encrypt.c (hunk 13 above)
Comment (quarium) [reply lib/upipe-dvbcsa/upipe_dvbcsa_encrypt.c]: You removed it from upipe_dvbcsa_encrypt =)

Comment (funman) [reply lib/upipe-dvbcsa/upipe_dvbcsa_encrypt.c]: No I used git mv ..

File: lib/upipe-dvbcsa/upipe_dvbcsa_encrypt.c (hunk 9 above)
Comment (funman) [reply lib/upipe-dvbcsa/upipe_dvbcsa_encrypt.c]: Done
